### ISSUE TITLE
API error on SWT on a 4.34 master

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-pde/eclipse.pde#1386]
I see API error on **SWT** if using https://download.eclipse.org/eclipse/downloads/drops4/S-4.33RC2a-202409030240/ as API baseline and https://download.eclipse.org/eclipse/downloads/drops4/I20240904-0530/ as IDE.

SWT error is about changed execution environment.

I've rebuilt everything few times, no changes, errors do not disappear.
![image](https://github.com/user-attachments/assets/ebe52849-ed3d-44c7-8102-763973e8e40c)

Note: originally reported API errors for JDT were gone after switching default workspace JDK from 21 to 17 and back, only SWT error remains.

**Comment by iloveeclipse** [comment]
Previous SWT related bug: https://github.com/eclipse-pde/eclipse.pde/issues/1283

**Comment by merks** [comment]
I felt like this was a ground hog day issue that was fixed but now its back.  😱   That's what happens when there is no test. 😭

**Comment by iloveeclipse** [comment]
Interestingly, I can't reproduce JDT errors after switching default workspace JDK to Java 17 version (I had 21 before and tried to change that because I've assued that would "mute" SWT errors) and back. But instead, JDT errors disappeared but SWT still has errors ???

**Comment by iloveeclipse** [comment]
I see in debugger in `org.eclipse.pde.api.tools.internal.model.BundleComponent.init()`, that the binary fragment has empty array as `fdeclaredRequiredEE`, but workspace one gets `JavaSE-17` (either current EE of the workspace or the one from host). So we have a mismatch... And yes, that looks like regression from https://github.com/eclipse-pde/eclipse.pde/pull/1302.

**Comment by merks** [comment]
FYI, I see both errors in my workspace:

![image](https://github.com/user-attachments/assets/a2f13e86-e58a-4416-9b18-8cc328b24500)

**Comment by iloveeclipse** [comment]
Which JDK is set as default in the workspace? Could you try to change that?0

**Comment by merks** [comment]
I have 17 as the default:

![image](https://github.com/user-attachments/assets/9a3294ea-f682-439a-95d6-ae9c076b06bd)

**Comment by iloveeclipse** [comment]
> have 17 as the default:

Could you just try to change it to 21, restart & rebuild JDT? That helped me, and after that I was able to switch back and still no errors.

**Comment by iloveeclipse** [comment]
I see contract violation in `org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getRequirements(String)`
It says:
`The returned list will be empty if this resource declares no requirements in the specified namespace.`
However it returns non empty list for SWT fragments because in org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getDeclaredRequirements(String) it calls org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getGenericRequires() that returns "Java 17".

Stack is
```
BundleDescriptionImpl.getDeclaredRequirements(String) line: 941	
BundleDescriptionImpl.getRequirements(String) line: 1275	
ManifestUtils.getRequiredExecutionEnvironments(Resource) line: 317	
ProjectComponent(BundleComponent).init() line: 317	
ProjectComponent(BundleComponent).getSymbolicName() line: 904	
WorkspaceBaseline(ApiBaseline).addComponent(IApiComponent) line: 326	
```

I can't say anything about that, it looks wrong, I will push a smaller workaround that fixes SWT error in a moment.

**Comment by iloveeclipse** [comment]
@merks : could you try https://github.com/eclipse-pde/eclipse.pde/pull/1387 for SWT error?

**Comment by merks** [comment]
> > have 17 as the default:
> 
> Could you just try to change it to 21, restart & rebuild JDT? That helped me, and after that I was able to switch back and still no errors.

The problem is there for 21 and when I switch back.

**Comment by merks** [comment]
> @merks : could you try #1387 for SWT error?

Sorry, I have to get ready to go out for dinner.  Maybe later when I'm back...

**Comment by HannesWell** [comment]
If I understand the issue correctly I think the new method in ManifestUtils should filter these artificial requirement out. That method is supposed to return what is declared in the manifest.

The test-case added in #1302 could also be extended to test the case if no EE is required.

I can look into both later this evening. 

IIRC I checked the code for these artificial EE requirements when reviewing #1302, but it was probably only for the `getExecutionEnvironments()` method.

**Comment by ptziegler** [comment]
This issue is effectively the inverse of what I ran into, based on the two possible configurations where an error may be shown:

- The baseline component _does_ specify an EE, but the workspace component _doesn't_.
- The baseline component _doesn't_ specify an EE, but the workspace component _does_.

My issue is related to the first case, where the problem lied in the workspace component specifying the EE via the `Require-Capability` header, which wasn't supported at that time and was therefore ignored.

The issue @iloveeclipse is having is related to the second case, where the gtk fragment _doesn't_ have an EE on the update site (and therefore in the baseline), but _does_ have it in the SWT workspace.

**Comment by iloveeclipse** [comment]
> gtk fragment _doesn't_ have an EE on the update site (and therefore in the baseline), but _does_ have it in the SWT workspace.

GTK  fragment does not have EE in the manifest!

**Comment by ptziegler** [comment]
> > gtk fragment _doesn't_ have an EE on the update site (and therefore in the baseline), but _does_ have it in the SWT workspace.
> 
> GTK fragment does not have EE in the manifest!

Then I must've overlooked it... You are correct, of course. I've created https://github.com/eclipse-pde/eclipse.pde/pull/1388 to handle this regression and also tested it against your PR, which makes the test succeed.

Though that must mean the ManifestUtils method somehow still adds those artificial requirements, even if they are not in the manifest.

**Comment by iloveeclipse** [comment]
> Though that must mean the ManifestUtils method somehow still adds those artificial requirements, even if they are not in the manifest.

See above: https://github.com/eclipse-pde/eclipse.pde/issues/1386#issuecomment-2329338273

**Comment by iloveeclipse** [comment]
> I can look into both later this evening.

Thanks, would be good to have the fix merged soon

**Comment by HannesWell** [comment]
> However it returns non empty list for SWT fragments because in org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getDeclaredRequirements(String) it calls org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getGenericRequires() that returns "Java 17".

I'm investigating why this is happening, but didn't found the cause yet.
I'll continue tomorrow so that this is resolved ASAP.

**Comment by HeikoKlare** [comment]
We would like to merge an SWT PR that will bump the minor version of the SWT bundles (https://github.com/eclipse-platform/eclipse.platform.swt/pull/1409)

From my understanding, that will "implicitly" resolve the discussed API errors. I would like to make sure that merging it will not make analysis of this issue more complicated. So do you think we can merge that PR or should we wait for more progress or resolution of this issue? That would, of course, also apply to any further SWT PR that might require a minor version bump.

**Comment by iloveeclipse** [comment]
> I would like to make sure that merging it will not make analysis of this issue more complicated.

I think the problem is understood and we have also a test case now via https://github.com/eclipse-pde/eclipse.pde/pull/1388.

> So do you think we can merge that PR or should we wait for more progress or resolution of this issue?

Feel free to merge SWT code that can be merged.

**Comment by merks** [comment]
FYI, I did a completely from scratch platform SDK setup and the JDT problem recurs there:

![image](https://github.com/user-attachments/assets/5a9ecd5b-6873-47b8-a304-f4dd4c3eb8d8)

But it doesn't occur from a debug launch.   😢

**Comment by merks** [comment]
In the debugger I see much more information about what the tool thinks has changed:

![image](https://github.com/user-attachments/assets/fec47c3c-2bdc-4aa7-b2f4-4d5405af5242)

But note that what is says has changed has not really changed.

**Comment by merks** [comment]
It looks like things are going wrong here

![image](https://github.com/user-attachments/assets/396aca01-61db-4699-ba68-938679e15511)

The first description is processed and it has the correct information, but then the second one is processed for the mock thing and that one is inappropriate.   I'm not sure if the order of these things is deterministic because the last one always wins.

And while exact has been set, but a wrapper object is returned and so nothing determines that the thing is exact.

![image](https://github.com/user-attachments/assets/88dc5a97-ed1d-4c13-8483-0e1c667f5d59)

How best to fix this...

**Comment by iloveeclipse** [comment]
Ed, do you mean, that analysis of the "mock" bundle makes the core bundle data corrupt, or the "mock" bundle analysis reports results to the wrong bundle? Why is "mock" thing analyzed at all?

**Comment by merks** [comment]
There are two descriptions and the last one wins.  So the mock bundle's description wins.  I assume everything in the workspace is analyzed.  That mock thing has an NonAPIProjectDescription, whatever that is.  Keep in mind I really don't know the design of the API tools.    I'm considering either how to make use of exact information better how to prioritize a ProjectAPIDescription's annotation over those of a NonAPIProjectDecription.  So I'm continuing to debug and try out solutions.

**Comment by iloveeclipse** [comment]
@merks : thanks to your great observation, I can reproduce too. 
Once I open `org.eclipse.jdt.core.tests.builder.mockcompiler` project & rebuild `org.eclipse.jdt.core` I see API error.

The  `org.eclipse.jdt.core.tests.builder.mockcompiler` project is a **fragment** of `org.eclipse.jdt.core` bundle.
But that is 1) a test and 2) why should it affect annotation visibility of the host types?

**Comment by merks** [comment]
I have prototyped what I believe is an appropriate fix so the method will be like this:
```
	public IApiAnnotations resolveAnnotations(IElementDescriptor element) {
		IApiAnnotations bestMatchAnnotation = null;
		for (IApiDescription fDescription : fDescriptions) {
			IApiAnnotations ann = fDescription.resolveAnnotations(element);
			if (ann != null) {
				bestMatchAnnotation = ann;
				if (ann.isExact()) {
					return ann; // if exact, return else keep looking for best match
				}
			}
		}
		return bestMatchAnnotation;
	}
```
I added a default isExact method to IApiAnnotations and make TypeAnnotations delegate to the IApiAnnotations that it wraps.

But I have to go out now and will create a PR later.

**Comment by merks** [comment]
There was a "going out" delays so now there is a PR being built.

**Comment by HannesWell** [comment]
> I see contract violation in `org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getRequirements(String)`
> It says:
> `The returned list will be empty if this resource declares no requirements in the specified namespace.`
> However it returns non empty list for SWT fragments because in org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getDeclaredRequirements(String) it calls org.eclipse.osgi.internal.resolver.BundleDescriptionImpl.getGenericRequires() that returns "Java 17".

OK, I found the cause for this and PDE itself is actually to blame, `BundleDescriptionImpl` is totally fine.
And actually not surprisingly the cause close to the place that as duplicated and worked-around in https://github.com/eclipse-pde/eclipse.pde/pull/1292: 
https://github.com/eclipse-pde/eclipse.pde/blob/0953a33f0825a551ba91983b9f80bb3a95ebf5e9/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/MinimalState.java#L141-L160

From the comments I believe we should really leave the injection of the synthetic EE in place. To fix this issue we should just restore the check for an explicit EE in `BundleComponent`, just like it was done in #1292 respectively restored in #1387.
I just created https://github.com/eclipse-pde/eclipse.pde/pull/1395, that restores this check but re-uses the method from `MinimalState`, which hopefully also creates a link in the code.

Alternatively we could also add another synthetic capability provided by a bundle with no EE and could check its existence.
This would also have the advantage that it could be checked in `ManifestUtils.getRequiredExecutionEnvironments()`. I wanted to use that method in more places in the future but I'm not yet sure if we actually always want only the declared EE requirements or sometimes also the 'effective' requirements. Probably the future has to show that.

  https://github.com/eclipse-pde/eclipse.pde/blob/0953a33f0825a551ba91983b9f80bb3a95ebf5e9/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/util/ManifestUtils.java#L316-L321.
